File: src/main/java/cc/sukazyo/cono/morny/bot/event/OnUserSlashAction.java
package cc.sukazyo.cono.morny.bot.event;

import cc.sukazyo.cono.morny.MornyCoeur;
import cc.sukazyo.cono.morny.bot.api.EventListener;
import cc.sukazyo.cono.morny.util.tgapi.TGToStringFromMessage;
import cc.sukazyo.untitled.util.command.CommonCommand;
import cc.sukazyo.untitled.util.string.StringArrays;

import com.pengrad.telegrambot.model.Message;
import com.pengrad.telegrambot.model.Update;
import com.pengrad.telegrambot.model.request.ParseMode;
import com.pengrad.telegrambot.request.SendMessage;

import javax.annotation.Nonnull;

import static cc.sukazyo.untitled.util.telegram.formatting.MsgEscape.escapeHtml;

public class OnUserSlashAction extends EventListener {
	
	@Override
	public boolean onMessage (@Nonnull Update event) {
		final String text = event.message().text();
		if (text == null) return false;
		
		if (text.startsWith("/"))
		{
			
			/// Due to @Lapis_Apple, we stopped slash action function at .DP7 groups.
			/// It may be enabled after some updates when the function will not be conflicted to other bots.
			// if (event.message().chat().id() == ) return false;
//{			if (event.message().chat().title() != null && event.message().chat().title().contains(".DP7")) {
//				logger.info(String.format("""
//					Chat slash action ignored due to the following keyword.
//					 - %s
//					 - ".DP7\"""",
//						TGToString.as(event.message().chat()).toStringFullNameId()
//				));
//				return false;
//			}
			
			final String[] action = CommonCommand.format(text);
			action[0] = action[0].substring(1);
			
			if (action[0].matches("^\\w+(@\\w+)?$")) {
				return false; // 忽略掉 Telegram 命令格式的输入
			} else if (action[0].contains("/")) {
				return false; // 忽略掉疑似目录格式的输入
			}
			
			final boolean isHardParse = "".equals(action[0]);
			/* 忽略空数据 */ if (isHardParse && action.length < 2) { return false; }
			final String verb = isHardParse ? action[1] : action[0];
			final boolean hasObject = action.length != (isHardParse?2:1);
			final String object =
					hasObject ?
					StringArrays.connectStringArray(action, " ", isHardParse?2:1, action.length-1) :
					"";
			final Message origin = event.message();
			final Message target = (event.message().replyToMessage() == null ? (
					origin
			): (
					event.message().replyToMessage()
			));
			
			MornyCoeur.extra().exec(new SendMessage(
					event.message().chat().id(),
					String.format(
							"%s %s%s %s %s!",
							TGToStringFromMessage.as(origin).getSenderFirstNameRefHtml(),
							escapeHtml(verb), escapeHtml((hasObject?"":"了")),
							origin==target ?
									"<a href='tg://user?id="+TGToStringFromMessage.as(target).getSenderId()+"'>自己</a>" :
							TGToStringFromMessage.as(target).getSenderFirstNameRefHtml(),
							escapeHtml(hasObject ? object+" " : "")
					)
			).parseMode(ParseMode.HTML).replyToMessageId(event.message().messageId()));
			
			return true;
			
		}
		return false;
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/bot/event/OnUserSlashAction.scala
package cc.sukazyo.cono.morny.bot.event

import cc.sukazyo.cono.morny.MornyCoeur
import cc.sukazyo.cono.morny.bot.api.EventListener
import cc.sukazyo.cono.morny.util.tgapi.formatting.TelegramFormatter.*
import cc.sukazyo.cono.morny.util.tgapi.formatting.TelegramParseEscape.escapeHtml as h
import cc.sukazyo.cono.morny.util.UniversalCommand
import com.pengrad.telegrambot.model.Update
import com.pengrad.telegrambot.model.request.ParseMode
import com.pengrad.telegrambot.request.SendMessage

import scala.language.postfixOps

object OnUserSlashAction extends EventListener {
	
	private val TG_FORMAT = "^\\w+(@\\w+)?$"r
	
	override def onMessage (using update: Update): Boolean = {
		
		val text = update.message.text
		if text == null then return false
		
		if (text startsWith "/") {
			
			val actions = UniversalCommand(text)
			actions(0) = actions(0) substring 1
			
			actions(0)
			
			actions(0) match
				case TG_FORMAT(_) =>
					return false
				case x if x contains "/" => return false
				case _ =>
			
			val isHardParse = actions(0) isBlank
			def hp_len(i: Int) = if isHardParse then i+1 else i
			if isHardParse && actions.length < 2 then return false
			val v_verb = actions(hp_len(0))
			val hasObject = actions.length != hp_len(1)
			val v_object =
				if hasObject then
					actions slice(hp_len(1), actions.length) mkString " "
				else ""
			val origin = update.message
			val target =
				if update.message.replyToMessage == null then
					origin
				else update.message.replyToMessage
			
			MornyCoeur.extra exec SendMessage(
				update.message.chat.id,
				"%s %s%s %s %s!".format(
					origin.sender_firstnameRefHTML,
					h(v_verb), if hasObject then "" else "了",
					if (origin == target)
						s"<a href='tg://user?id=${origin.sender_id}'>自己</a>"
					else origin.sender_firstnameRefHTML,
					if hasObject then h(v_object+" ") else ""
				)
			).parseMode(ParseMode HTML).replyToMessageId(update.message.messageId)
			true
			
		} else false
		
	}
	
}
